### ISSUE TITLE
add regression capabilities

### ISSUE THREAD

**Issue by jovo** [neurodata/progressive-learning#5]

**Comment by latasianguy** [comment]
Can I be assigned to this issue?

**Comment by latasianguy** [comment]
@levinwil @jdey4 should I be writing a new forest class for this, along with its deciders/transformers/voters? I think once I do that, I can find some regression dataset benchmarks to test the implementation in a notebook

**Comment by levinwil** [comment]
@latasianguy I will mention that this is a very difficult and tricky issue. 

In terms of implementation, we already have code that implements regression that is clearly the same algorithm as in classification, just modified to be regression instead. That code currently exists on the regression branch of this repository, both for networks and forests. 

However, code is just language you use to communicate an algorithm to the computer. The difficulty comes in with this issue in figuring out what that algorithm should be. Regression is hit very hard by adversarial transformers, in terms of transfer efficiency. Typically, we would hope (and in classification, we do see this) that adding more transformers results in a backward transfer efficiency >= 1. However, in regression, my impression is that the algorithms that have been tried and implemented have not always resulted in backward transfer efficiency >= 1. I have had several discussions with those who previously attempted to come up with various deciders (and voters, but more emphasis on deciders) to alleviate this problem, and some things have worked for certain datasets. So there is hope. But I wanted to mention that this is not a straight forward "we have this algorithm and I need to implement it and then it will work just as well as classification." We already have that and it exists on the regression branch. This issue will primarily be coming up with new algorithms (most likely voters and deciders) to alleviate adversarial transformers. 

If you like designing algorithms, then great! But I wanted to give a full disclosure about what this issue entails.

**Comment by latasianguy** [comment]
@levinwil are there any papers you would recommend for me to get a scope of this problem?

**Comment by levinwil** [comment]
@latasianguy I personally don't know of any papers that directly address the problems we have been experiencing. I can provide some papers that are related to potential solutions I think might be worth trying. Would that be useful? It's up to you how you'd like to spend your time, and if there's a chance I send you some papers and you spend days or weeks reading them, implement them in our framework, and none work. 

@jovo Might have some papers that he thinks are useful. 

I will also mention that, in general, the regression landscape is very sparse and there isn't much to read (not to mention, most of the stuff isn't great). I can go into my impression of why that is, but I don't think that's particular important - I just wanted to mention that it might be difficult to find papers that directly address this issue.

**Comment by latasianguy** [comment]
@levinwil I think that would be useful, it'll be a good experience for me working on this type of open-ended problem that may or may not have a good solution

**Comment by levinwil** [comment]
@latasianguy 

**KNN Background**
1. Have you read the original Stone's consistency theorem of KNN regression paper? If not, it's fine. It's a very technical and dense paper to convey a simple point (and this is an oversimplified interpretation): as a KNN regressor trains on increasingly more IID data from the distribution it'll see at inference time, the test error (measured through, I believe, mean absolute error) approaches 0. This just means it's a consistent estimator. 

2. KNN is great. But it's hit hard by the curse of dimensionality. Here is some background on why (and some more useful background): https://www.cs.cornell.edu/courses/cs4780/2018fa/lectures/lecturenote02_kNN.html

3. Here is a post about one way to alleviate the curse of dimensionality when using KNN: https://stackoverflow.com/questions/5751114/nearest-neighbors-in-high-dimensional-data. It mentions one way of alleviating the curse of dimensionality: Locally Sensitive Hashing (LSH). There may be other ways, I'm not sure. This is just my go-to.

4. Here's a post that explains LSH in a way that I think is simple: https://stackoverflow.com/questions/12952729/how-to-understand-locality-sensitive-hashing. There are plenty of other resources which likely explain it well. 

**Selfish Background That Describes a Method I Want to Try Since I Developed The Idea**
1. I wrote an algorithm a few months ago that I believe directly addresses adversarial task alleviation, and in a local way. I have a conjecture (and a pretty strong one at that but could be wrong) that it will alleviate adversarial tasks. Maybe there exists other methods that alleviate it even more, and maybe my conjecture is wrong. But I selfishly want to try this idea, because everyone thinks there ideas are going to work haha [NICE.pdf](https://github.com/neurodata/progressive-learning/files/5278529/NICE.pdf)

**Comment by latasianguy** [comment]
@levinwil I think I want to play around with running regression forests with normal data and adverserial data. I'm working on the branch for regression. Will a toy example of a underlying polynomial curve like Ax^2 + Bx + C and a higher order polynomial, both with added noise, work to simulate a two task regression problem? This way I can understand the transfer efficiencies and play around with some parameters. Let me know if there is a better way to create toy datasets

**Comment by levinwil** [comment]
@latasianguy That is along the lines of what MSR people and I discussed. That was how we started (thinking of a simple curve and a higher order polynomial). Just a quick reminder: an adversarial transformer with respect to task A transforms the data such that - when a voter is trained on top of task A data transformed with the adversarial transformer - BTE < 1 with respect to task A (i.e. that voter harms the overall voter ensemble). This will be most apparent when there is high loss on that voter that is trained on task A data that went through the adversarial transformer. 

This is where we ended up in terms of setting up toy adversarial experiments. For all of the below experiments, we care about Task A, which is a simple linear function of y = x, with very high sample size and very low noise (make standard deviation = 0.1 or less). This can be easily made with the following function: https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_regression.html. Here are the following Task B settings we were thinking of making: 

1. (simplest and most unrealistic) make a new transformer class that outputs random partitions. That should results in data that is pretty much random, and the voter that is trained on top of that data should result in pretty high loss, and be unhelpful with respect to BTE for previous tasks. Thus, to simulate an adversarial transformer, simply make this random transformer the transformer for task B.
2. (slightly more realistic) the data used to train the task B transformer is the line y = 0. The partitions, therefore, should be pretty much random since it doesn't make a difference where they lie: every leaf is automatically pure. 
3. (most realistic) the data used to train the task B transformer is the line y = x, but with a VERY low sample size and MUCH more noise (e.g. make standard deviation = 2.0 or more). You can use the same [function](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_regression.html) to make this as to make the task A data.

Keep in mind that the steps for training the PL should be as follows for each setting:
1. use add_task on the Task A data 
2. use predict to get task A test predictions 
3. compute test loss and save it 
4. use add_transformer on the task B adversarial transformer (it will train everything internally)
5. use predict to get task A test predictions 
6. compute test loss and save it 
7. compare #3 and #6. In the presence of an adversarial transformer, #6 will likely be much higher than #3. We want to lower that difference as much as possible.

**Comment by levinwil** [comment]
@latasianguy Do you think you will make functions that generate the 3 mentioned settings? If so, do you have any idea when you will get around to doing so? MSR people are interested in this stuff, so it's fine if you aren't going to get to it anytime soon - but please let me know so I can do it.

**Comment by latasianguy** [comment]
@levinwil I can get try to start it this week, but I'll have to play around with the regression branch since I haven't run any experiments before

**Comment by latasianguy** [comment]
@levinwil I'm trying to get setting 2 working in a notebook right now. I'm using a simple least square loss to compute the test loss and was wondering what I should be iterating over. In other experiments, like the XOR-NXOR, we iterated over an increasing sample size, would the same apply here?

**Comment by levinwil** [comment]
@latasianguy 
1. Settings 2 and 3 do not need explicit adversarial transformers. You should be able to simply train a TreeRegressionTransformer (via add_transformer) and that transformer should be unhelpful/adversarial.

2. For now, let’s iterate over the max_depth of TreeRegressionTransformer, and make it such that the data fed to this transformer is the line y=0 with some added noise (as opposed to to no noise). We should see that increasing the max_depth should produce finer-grain partitions, and therefore be more hopeful. But that is just a conjecture. 

3. As for the TreeRegressionRandomDecider, that is exactly correct: just output a random integer. Let’s assume it has a parameter for max_depth. Then there would be 2**(max_depth - 1) leaves, so that should be the number of possible partitions it can output. I.e. it should output ‘return np.random.randint(low=0, high=2**(max_depth - 1), size = len(X))’

**Comment by latasianguy** [comment]
@levinwil I got a framework working but the generalization errors I'm getting don't seem to be matching up. Am I using the add_transformer/predict function incorrectly or is my generated too close together? I've linked the [notebook](https://github.com/latasianguy/progressive-learning/blob/adverserial_regression/tutorials/TreeRegressionExample.ipynb).

**Comment by levinwil** [comment]
I believe this is because you are setting max_depth so high for both transformers. I would try the following:
1. set max_depth of transformer_A to be 30
2. perform the same exact experiment, but when you iterate across max_depth, only modify the max_depth of transformer_B, and start at a max_depth of 1 and increase to 30

**Comment by levinwil** [comment]
@latasianguy I would also try plotting a third line and that is the performance when you just send inference examples through transformer 1 (setting transformer_ids = [1] in predict)

**Comment by latasianguy** [comment]
@levinwil the graph I've been getting has the mean error for the normal and adversarial transformer (transformer 0 + 1) and the mean error for only the adversarial transformer as the same (transformer 1). Even though transformer 1 is trained on useless data, I don't think that the predictions would be the same, right? [Notebook](https://github.com/latasianguy/progressive-learning/blob/adverserial_regression/tutorials/TreeRegressionExample.ipynb)

**Comment by levinwil** [comment]
@latasianguy this is all because you’re using KNNRegresaionDecider. As mentioned, most of the work here will be on chancing the decider. We want to first highlight that using a SimpleAverage decider is not effective enough. Please use SimpleAverageDecider. Furthermore - and this part is crucial - you must use ‘predict_proba’ instead of predict when running inference on the progressive learner. I urge you to look at the actual functions in the SimpleAverageDecider to see why - ‘predict’ assumes you return the highest probability class, which is not what we want.

**Comment by levinwil** [comment]
@latasianguy if I were you, I’d also confirm in some way that your task B data has outputs of 0. You could’ve made it much simpler by saying taskB_y = np.zeros((len(taskB_X))) and that’ll certainly make sure your labels for task B are zero. If the task B labels aren’t 0 exactly then the whole experiment is faulty.

**Comment by latasianguy** [comment]
@levinwil from the feedback on Wednesday, I think I'll do the following:

- run experiments with task A transformer at depth=30, task B transformer going from depths=1 - 30, stepsize=1
- run experiments with task A transformer with tree=25, task B transformer trees=5 - 50, stepsize=5
- run experiments with 1 task A transformer, and 1-30 task B transformers

I'll plot the mean generalization error for the both of them, let me know if any of this should be adjusted

**Comment by levinwil** [comment]
@latasianguy I would apply noise after normalization to the taskB dataset. So instead of doing:

y_new = (y_new - regressor.intercept_) / regressor.coef_
y_new = y_new.reshape(X_new.shape[0], -1)

Try 

y_new = (y_new - regressor.intercept_) / regressor.coef_
y_new = y_new.reshape(X_new.shape[0], -1)
y_new += np.random.normal(size = y_new.shape, scale = 10.0)

I would also sample 'int(sample_size / 10)' points from taskB instead of 'int(ceil(np.sqrt(sample_size)))' points from taskB

Does that make sense? 

These are extreme example, but we're just trying to illustrate a point here - it doesn't have to be realistic

**Comment by latasianguy** [comment]
@levinwil I tried what you suggested and updated the [notebook](https://github.com/latasianguy/progressive-learning/blob/adverserial_regression/tutorials/TreeRegressionExample.ipynb) but it didn't seem to work (graph at top right), I scaled the noise up to 20.0 and it is still showing that the normal transformer's generalization error is higher

also, I added a RandomTransformer class to transformers.py using the TreeRegressionTransformer code but when I try to import it, it gives me an import error. Is there a step I have to do first before trying to import the new transformer?

**Comment by levinwil** [comment]
I have no idea what the distinction between these figures are. I have no idea what most of this is. Could you please explain it more so I can help? I highly doubt that a more noisy setting with less data outperforms a noiseless setting with more data. I imagine you have a bug somewhere, so if I were you I’d make sure absolutely with no certainty that you are providing the correct experiment parameters. Machine learning says performance decreases as noise increases and sample size decreases, so that setting should produce worse results. 

As far as the random transformer, if you change the proglearn package locally and want to import that change then you have to reinstall the package with “python setup.py install” from the root directory of the project

**Comment by latasianguy** [comment]
@levinwil ![image](https://user-images.githubusercontent.com/54046695/95229623-c8e46200-07ce-11eb-9225-9ccfe936e937.png)
The figures on the left column are the experiments done on the adversarial transformer trained on y=0 data.
The figures on the right column are the experiments done on the adversarial transformer trained on noisy y=x data.
The figures along the top row are running the adversarial transformer from different depths with fixed trees.
The figures along the bottom row are running the adversarial transformer from different number of trees with fixed depth.

I'll make the figures better today.

**Comment by levinwil** [comment]
I don't understand why on the top right, the Adversarial wins while on the bottom right, Adversarial Loses. Aren't they the same setting? What is the fixed number of trees in the top right and what is the fixed depth in the bottom right?

**Comment by levinwil** [comment]
The top right and bottom right should have one overlapping experiment where they have the same number of trees and max_depth. I think your depth in the bottom right is 30, and your n_trees in the top right is 25. So, the column in the top right at depth= 30 (with fixed 25 trees) should be the same as the column in the bottom right at trees = 25 (with fixed 30 depth). Those have the same hyperparams and the same data settings, so the results should be the same. But they're not. So something is wrong. Those 2 settings should be equivalent, unless I'm missing something.

**Comment by levinwil** [comment]
Everything looks about correct, other than the bottom right. I'd investigate there and make sure your experiment parameters are being fed correctly. A good check is that the columns mentioned in the above comment should be about the same. Does that make sense?

**Comment by latasianguy** [comment]
@levinwil yep, that sounds good, I'll look over everything again

**Comment by latasianguy** [comment]
@jovo @levinwil here's a plan I made from the feedback I got from Wednesday

- [ ] Replicate XOR_NXOR experiment (theres's already a [tutorial](https://github.com/neurodata/ProgLearn/blob/docs/docs/tutorials/xor_nxor_exp.ipynb))
- [ ] Use the regression branch of the repo to test a simple regression extension, except with y=x as the data, instead of XOR_NXOR. Since this is regression, I'm assuming that we can use two progressive learners trained on y=x tasks with varying noise to simulate the XOR and NXOR learner. This experiment should show BTE > 1 with respect to the first task.
- [ ] Add more tasks, which should all show BTE w.r.t. the first task.
- [ ] Extend to only transformers if the above work?

Let me know if I'm understanding anything wrong, thanks!

**Comment by levinwil** [comment]
For the second point, I don’t understand. Why would you use two progressive learners on y=x? Do you mean to say you’ll use one progressive learner but two transformers? If so, it should be fine to just use y=x with the same amount of noise (although I personally think it would be better to use y=x for task 1 and y=-x for task 2, as that more directly reflects XOR/NXOR in that the second task is rotated 90 degrees from task 1). 

The above experiment should show BTE >1 with respect to tack 1 AND FTE > 1 with respect to task 2.

As for your third point, if you only care about BTE with respect to task 1, using add_task and add_transformer will have identical effects on BTE when adding any task past task 1 to the progressive learner - there is no need to use add_task for those future tasks, as add_task will train a bunch of unnecessary voters. Feel free to use add_task for tasks past task 1 but it’ll just slow you down.

**Comment by latasianguy** [comment]
@levinwil Updated with a new [notebook](https://github.com/latasianguy/progressive-learning/blob/adverserial_regression/tutorials/regression_exp.ipynb). Here are the graphs from the notebook:
![image](https://user-images.githubusercontent.com/54046695/95765262-9aa7cc00-0c7f-11eb-94da-db67bb91d3a0.png)
![image](https://user-images.githubusercontent.com/54046695/95765276-9e3b5300-0c7f-11eb-94e3-cefcf276d281.png)

I have a few questions:
1. The BTE looks to have a steep drop, will increasing the sample size for the y=-x transformer help?
2. Two task classification errors have a random one that is pure chance = 0.5, regression using MSE doesn't, can I just use error=1 as a placeholder when training with the single task for the first 700 samples?

**Comment by latasianguy** [comment]
**Normal Regression Experiment**
![image](https://user-images.githubusercontent.com/54046695/95915778-ad9cc800-0d75-11eb-8f5d-8c87fe919ccb.png)
![image](https://user-images.githubusercontent.com/54046695/95915808-bab9b700-0d75-11eb-8f12-c64f8912a39e.png)

**Comment by latasianguy** [comment]
**Adversarial Experiment**
![image](https://user-images.githubusercontent.com/54046695/95915922-e177ed80-0d75-11eb-8b19-aa16c1fdcbdd.png)
![image](https://user-images.githubusercontent.com/54046695/95915936-e89efb80-0d75-11eb-9efc-65f428ad9266.png)

**Comment by latasianguy** [comment]
@levinwil so from Wed, I guess I should start by making function that generate two 3D gaussians (y, x1, x2) for task 1 and task 2, mess around with their covariance matrices to give them the same distribution as XOR, NOR (I think jovo mentioned correlations of 0.5 and -0.5) and then run the same experiments without a max depth and see how the BTE and FTE look like? should I also plot how the partitions are generated? if so, is there a link to similar experiments that have done that?

**Comment by levinwil** [comment]
@latasianguy That all sounds good to me. No need to plot the partitions, though. Maybe we can do that later, but that's not particularly important right now.

**Comment by latasianguy** [comment]
@levinwil New notebook linked [here](https://github.com/latasianguy/progressive-learning/blob/adverserial_regression/tutorials/regression_exp_gaussian.ipynb). I didn't set a max depth so I was wondering if it overfit to the task 1 Gaussian data since the generalization error for task 2 is rising so the partitions learnt is less helpful as more task 2 data is added?

**Comment by levinwil** [comment]
@latasianguy Could you please try setting max_depth = 10? If you set max_depth = None, that means the transformers will train to purity and then more data means more overfit partitions. Does that make sense?
